torchdata/stateful_dataloader/incremental_state.py

```diff
@@ -0,0 +1,178 @@
+                )
+                data.update(flat)
+        else:
+            if len(prefix) > 0:
+                data[prefix] = nested_data
+            else:
+                data = nested_data
+        return data
+
+    @classmethod
+    def unflatten(cls, flat_data: Optional[Dict[str, Any]], separator: str = "/"):
+        if flat_data is None:
+            return None
+        nested_data = {}
+        for key, value in flat_data.items():
+            _keytoks = key.split(separator)
+            prefix = _keytoks[0]
+            suffix = separator.join(_keytoks[1:])
+            if len(_keytoks) == 1:
+                nested_data[prefix] = value
+                continue
+            if prefix not in nested_data:
+                nested_data[prefix] = {}
+            nested_data[prefix][suffix] = value
+        # now go through nested_data and combine them all
+        for k, v in nested_data.items():
+            if isinstance(v, dict):
+                nested_data[k] = cls.unflatten(v, separator=separator)
+        return nested_data
+
+
+class DeletionTombStone:
+    pass
+
+
+class IncrementalState:
+    def __init__(self, initial_state: Optional[Dict[str, Any]]):
+        self.flat_state = Flattener.flatten(initial_state)
+
+    def generate_delta(self, new_state: Dict[str, Any]):
+        new_flat_state = Flattener.flatten(new_state)
+        delta_flat_state = {}
+        all_keys = set()
+        if self.flat_state:
+            all_keys = set(self.flat_state.keys())
+        all_keys = all_keys.union(new_flat_state.keys())
+
+        for key in all_keys:
+            if self.flat_state is None or key not in self.flat_state:
+                # New key, retain it
+                delta_flat_state[key] = new_flat_state[key]
+                continue
+
+            if key not in new_flat_state:
+                # Key deletion, put in a tombstone
+                delta_flat_state[key] = DeletionTombStone()
+                continue
+
+            prev_value, new_value = self.flat_state[key], new_flat_state[key]
+            try:
+                if isinstance(prev_value, torch.Tensor) and isinstance(new_value, torch.Tensor):
+                    if torch.allclose(prev_value, new_value):
+                        continue
+                elif prev_value == new_value:
+                    continue
+            except Exception:
+                # Fallback to retaining new key/value
+                pass
+            delta_flat_state[key] = new_value
+        # Update internal state to the new state
+        self.flat_state = new_flat_state
+        return delta_flat_state
+
+    def apply_delta(self, flat_delta_state: Dict[str, Any]) -> None:
+        for key, update in flat_delta_state.items():
+            if self.flat_state is None:
+                self.flat_state = {}
+
+            if isinstance(update, DeletionTombStone):
+                # Remove key if present in the state
+                self.flat_state.pop(key, None)
```

Comment:
Oh I see, I think this will drop the entire branch so no need to manually delete everything. Wondering how that case actually plays out though

Reply:
Can you elaborate on this a bit more - entire branch as in the child nodes in a parent dict? The state in this function is always maintained as a flat_dict. The delta state generated so contains flat keys. So deletion iterates through all the flat keys and deletes it from the flat_state. 

Are you suggesting that we don't have to delete all these keys instead just have a marked for the top level key deletion and then delete all the child keys based on it?